### PR TITLE
Test：Add corner cases using sync point

### PR DESCRIPTION
**Pull request by sherry-1001**
### What is changed and how it works?
Add some test corner cases.
related PR: [168](https://github.com/pmem/kvdk/pull/168)

Tests 

- Unit test
   test.cpp:  `*SyncPoint*`
- Integration test
- No code

Side effects

- Performance regression
    - Consumes more CPU
    - Consumes more MEM
    - Consumes more PMEM

### CODE REVIEW

File: tests/tests.cpp [hunk 1]
```diff
@@ -1522,6 +1523,312 @@ TEST_F(EngineBasicTest, TestSortedCustomCompareFunction) {
...
+      for (int i = 0; i < cnt; ++i) {
+        std::string key = "key" + std::to_string(i);
+        std::string val = "val" + std::to_string(i);
+        wb.Put(key, val);
+      }
+      ASSERT_EQ(engine->BatchWrite(wb), Status::Ok);
+    });
+
+    ts.emplace_back([&]() {
+      TEST_SYNC_POINT("Test::Set::Start::0");
```
Comment (ZiyanShi) [inline tests/tests.cpp:1578]: Please give these Sync Points a meaningful name.
like "Test::before-set-k10v17", "Test::after-set-k8-v15". This can improve readability.

File: tests/tests.cpp [hunk 2]
```diff
@@ -1522,6 +1523,312 @@ TEST_F(EngineBasicTest, TestSortedCustomCompareFunction) {
...
+        {{"Test::Set::Finish::1",
+          "KVEngine::BatchWrite::AllocateRecord::After"}});
+    SyncPoint::GetInstance()->SetCallBack(
+        "KVEngine::BatchWrite::AllocateRecord::After", [&](void *size) {
+          size_t bsize = *((size_t *)size);
+          ASSERT_EQ(cnt, bsize);
+        });
+    SyncPoint::GetInstance()->SetCallBack(
+        "KVEngine::BatchWrite::Pesistent::0",
+        [&](void *) { write_batch.store(true); });
```
Comment (ZiyanShi) [inline tests/tests.cpp:1550]: Maybe this is some kind of thing that can be done by LoadDependency()?

File: engine/kv_engine.cpp [hunk 3]
```diff
@@ -1156,11 +1164,13 @@ Status KVEngine::BatchWrite(const WriteBatch &write_batch) {
       thread_res_[write_thread.id].persisted_pending_batch,
       space_entry_offsets);
 
+  std::vector<Status> statuss(write_batch.Size());
```
Comment (ZiyanShi) [inline engine/kv_engine.cpp:1167]: What's this?

File: engine/kv_engine.cpp [hunk 4]
```diff
@@ -1156,11 +1164,13 @@ Status KVEngine::BatchWrite(const WriteBatch &write_batch) {
...
   // Do batch writes
   for (size_t i = 0; i < write_batch.Size(); i++) {
     if (write_batch.kvs[i].type == StringDataRecord ||
         write_batch.kvs[i].type == StringDeleteRecord) {
       s = StringBatchWriteImpl(write_batch.kvs[i], batch_hints[i]);
+      TEST_SYNC_POINT_CALLBACK("KVEnigne::BatchWrite::BatchWriteRecord", &i);
```
Comment (ZiyanShi) [inline engine/kv_engine.cpp:1173]: Like "KVEnigne::BatchWrite::HasPersistedNthKV" ?

File: engine/dlinked_list.hpp [hunk 5]
```diff
@@ -359,6 +359,7 @@ class DLinkedList {
 
     iter_prev->next = offset;
     pmem_persist(&iter_prev->next, sizeof(PMemOffsetType));
+    TEST_SYNC_POINT("KVEngine::DLinkedList::emplaceBetween::PersistPrev");
```
Comment (ZiyanShi) [inline engine/dlinked_list.hpp:362]: "KVEngine::DLinkedList::emplaceBetween::AfterPrevLinkThisBeforeNextLinkThis"
This name is long, but somewhat suitable.

File: engine/kv_engine.cpp [hunk 6]
```diff
@@ -956,6 +959,7 @@ Status KVEngine::SSetImpl(Skiplist *skiplist, const StringView &user_key,
                             new_record_pmem_ptr, HashOffsetType::DLRecord);
       }
       purgeAndFree(existing_record);
+      TEST_SYNC_POINT("KVEngine::SSetImpl::Update::Finish");
```
Comment (ZiyanShi) [inline engine/kv_engine.cpp:962]: "KVEngine::SSetImpl::AfterUpdate"?

File: engine/kv_engine.cpp [hunk 7]
```diff
@@ -1138,14 +1143,17 @@ Status KVEngine::BatchWrite(const WriteBatch &write_batch) {
     spins_to_lock.emplace(batch_hints[i].hash_hint.spin);
   }
 
+  size_t batch_size = write_batch.Size();
+  TEST_SYNC_POINT_CALLBACK("KVEngine::BatchWrite::AllocateRecord::After",
+                           &batch_size);
```
Comment (ZiyanShi) [inline engine/kv_engine.cpp:1148]: "AfterAcquireAllSpins"

File: engine/kv_engine.cpp [hunk 8]
```diff
@@ -1217,6 +1227,7 @@ Status KVEngine::StringBatchWriteImpl(const WriteBatch::KV &kv,
     void *block_base =
         pmem_allocator_->offset2addr(batch_hint.allocated_space.offset);
 
+    TEST_SYNC_POINT("KVEngine::BatchWrite::Pesistent::0");
```
Comment (ZiyanShi) [inline engine/kv_engine.cpp:1230]: "KVEnigne::BatchWrite::BeforePersistNthKV",  consider put it in BatchWrite

File: engine/kv_engine.cpp [hunk 9]
```diff
@@ -1228,6 +1239,7 @@ Status KVEngine::StringBatchWriteImpl(const WriteBatch::KV &kv,
       std::abort();
     }
 
+    TEST_SYNC_POINT("KVEngine::StringBatchWriteImpl::HashInsertBefore");
```
Comment (ZiyanShi) [inline engine/kv_engine.cpp:1242]: "StringBatchWriteImpl::AfterPersistBeforeInsertInHashTable"

File: engine/kv_engine.cpp [hunk 10]
```diff
@@ -1272,6 +1285,7 @@ Status KVEngine::StringDeleteImpl(const StringView &key) {
     case Status::Ok:
       assert(entry_ptr->header.status == HashEntryStatus::Updating);
       purgeAndFree(hash_entry.index.string_record);
+      TEST_SYNC_POINT("DataEntry::StringDeleteImpl::purgeAndFree");
```
Comment (ZiyanShi) [inline engine/kv_engine.cpp:1288]: "StringDeleteImpl::AfterDeletion"
The spin is guarded. There's no way other thread can interrupt next step.

File: engine/kv_engine.cpp [hunk 11]
```diff
@@ -1314,12 +1328,15 @@ Status KVEngine::StringSetImpl(const StringView &key, const StringView &value) {
     kvdk_assert(!found || new_ts > data_entry.meta.timestamp,
                 "old record has newer timestamp!");
 
+    TEST_SYNC_POINT("KVEngine::StringSetImpl::PersistBefore");
```
Comment (ZiyanShi) [inline engine/kv_engine.cpp:1331]: "StringSetImpl::BeforePersistNewRecord"

File: engine/kv_engine.cpp [hunk 12]
```diff
@@ -1314,12 +1328,15 @@ Status KVEngine::StringSetImpl(const StringView &key, const StringView &value) {
...
     StringRecord::PersistStringRecord(block_base, sized_space_entry.size,
                                       new_ts, StringDataRecord, key, value);
 
+    TEST_SYNC_POINT("KVEngine::StringSetImpl::HashInsertBefore");
```
Comment (ZiyanShi) [inline engine/kv_engine.cpp:1335]: "AfterPersistBeforeInsertInHashTable"

File: engine/skiplist.cpp [hunk 13]
```diff
@@ -353,6 +356,8 @@ bool Skiplist::Insert(const StringView &key, const StringView &value,
   // create dram node for new record
   auto height = Skiplist::RandomHeight();
   if (height > 0) {
+    TEST_SYNC_POINT("KVEngine::Skiplist::Insert::CreateNode::Before" +
```
Comment (ZiyanShi) [inline engine/skiplist.cpp:359]: Let's discuss for a naming standard for all these Sync Points.

File: engine/kv_engine.cpp (hunk 9 above)
Comment (ZiyanShi) [reply engine/kv_engine.cpp:1242]: Do we need a Sync Point here? everything is locked?

**Review by ZiyanShi — APPROVED** [review]

File: tests/tests.cpp [hunk 14]
```diff
@@ -1520,6 +1521,384 @@ TEST_F(EngineBasicTest, TestSortedCustomCompareFunction) {
...
+// Then Repair
+
+TEST_F(EngineBasicTest, TestSortedRecoverySyncPointCaseTwo) {
+  Configs test_config = configs;
+  test_config.max_write_threads = 16;
+  ASSERT_EQ(Engine::Open(db_path.c_str(), &engine, test_config, stdout),
+            Status::Ok);
+  SyncPoint::GetInstance()->DisableProcessing();
+  SyncPoint::GetInstance()->Reset();
+  std::atomic<bool> visited(false);
```
Comment (ZiyanShi) [inline tests/tests.cpp:1785]: Is this variable necessary?

File: tests/tests.cpp [hunk 15]
```diff
@@ -1520,6 +1521,384 @@ TEST_F(EngineBasicTest, TestSortedCustomCompareFunction) {
...
+      }
+    }
+  }
+}
+
+// Example Case Two:
+//         A <-> C <-> D
+// Insert order: A, D, C
+// Delete:  A ----------> D
+//          A <- C <->D
```
Comment (ZiyanShi) [inline tests/tests.cpp:1775]: A<---------D
A<==>C-->D

File: tests/tests.cpp [hunk 16]
```diff
@@ -1520,6 +1521,384 @@ TEST_F(EngineBasicTest, TestSortedCustomCompareFunction) {
...
+          } else
+            ASSERT_EQ(got_val, "val" + std::to_string(i));
+        } else {
+          ASSERT_EQ(s, Status::NotFound);
+        }
+      }
+    }
+  }
+}
+
```
Comment (ZiyanShi) [inline tests/tests.cpp:1770]: Delete C, crash half way, now the state is ...

**Review by ZiyanShi — CHANGES_REQUESTED** [review]

**Review by ZiyanShi — APPROVED** [review]

**Review by JiayuZzz — APPROVED** [review]
